gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+            back_inserter(unrestricted));
+        return unrestricted;
+    }
+
+    auto propagate_using_smart_str(const vector<IntegerVariableID> & selectors, const vector<IntegerVariableID> & vars,
+        const SmartTuples & tuples, const vector<SmartTable::Forest> & forests, State & state) -> Inference
+    {
+
+        bool changed = false, contradiction = false;
+
+        VariableDomainMap unsupported{};
+        // Initialise unsupported values to everything in each variable's current domain.
+        for (const auto & var : vars) {
+            state.for_each_value(var, [&](Integer value) {
+                if (! unsupported.contains(var)) {
+                    unsupported[var] = vector<Integer>{};
+                }
+                unsupported[var].emplace_back(value);
+            });
+        }
+
+        // Check that feasible tuples are still feasible
+        // and also have them remove values from "unsupported" that they support
+        for (unsigned int tuple_idx = 0; tuple_idx < tuples.size(); ++tuple_idx) {
+            // Skip infeasible tuple
+            if (state.optional_single_value(selectors[tuple_idx]) == 0_i) {
+                continue;
+            }
+
+            for (const auto & tree : forests[tuple_idx]) {
+                // Initialise supported by tree to current variable domains
+                VariableDomainMap supported_by_tree{};
+
+                for (const auto & var : vars) {
+                    state.for_each_value(var, [&](Integer value) {
+                        if (! supported_by_tree.contains(var)) {
+                            supported_by_tree[var] = vector<Integer>{};
+                        }
+                        supported_by_tree[var].emplace_back(value);
+                    });
+                }
+
+                // First pass of filtering supported_by_tree and check of validity
+                if (! filter_and_check_valid(tree, supported_by_tree)) {
+                    // Not feasible
+                    switch (state.infer_equal(selectors[tuple_idx], 0_i, NoJustificationNeeded{})) {
+                    case Inference::NoChange: break;
+                    case Inference::Change: changed = true; break;
+                    case Inference::Contradiction: contradiction = true; break;
+                    }
+                    break;
+                }
+
+                if (contradiction) {
+                    return Inference::Contradiction;
+                }
+
+                filter_again_and_remove_supported(tree, supported_by_tree, unsupported);
+            }
+
+            if (state.optional_single_value(selectors[tuple_idx]) != 0_i) {
+                const auto unrestricted = get_unrestricted(vars, tuples[tuple_idx]);
+                for (const auto & var : unrestricted) {
+                    unsupported[var] = vector<Integer>{};
+                }
+            }
+        }
+
+        bool some_tuple_still_feasible = false;
```

Comment:
This feels potentially slow if there are lots of tuples and nearly all of them are infeasible. With Table and its use of an integer variable, we get fast iteration over sparse sets for free (allegedly). Perhaps this is worth considering?